README.md
Nginx statistic module
===================
[![Build Status](https://travis-ci.org/shoonoise/lua-nginx-statistics.svg?branch=master)](https://travis-ci.org/shoonoise/lua-nginx-statistics)
## Overview

Nginx Lua module to collect and show statistics.
See `nginx.conf` for example.

`collect_statuses.lua` module collect statistics for requests across location with `log_by_lua_file 'collect_stats.lua';` directive.

`show_stat.lua` response for JSON reply.

## Try it

For a start, you need nginx with [HttpLuaModule](http://wiki.nginx.org/HttpLuaModule), and lua json lib installed (`liblua5.1-json` for ubuntu).

* Clone this repository
* Copy \*.lua files from root to some path(`/usr/share/nginx/`, in example below)

Update your `nginx.conf`:

* Add `lua_shared_dict stat_dict 1M;` in http section
* Add `lua_package_path '/usr/share/nginx/?.lua;;';` in http section
* Add `log_by_lua_file '/usr/share/nginx/collect_stats.lua';` directive in location from which to you want collect statistic
* Add new location to get statistic and put `/usr/share/nginx/content_by_lua_file 'show_stat.lua'` directive inside
* Enjoy!

## Try it with docker

### Build it from Dockerfiles

* Install [Docker](https://www.docker.io/)
* clone this repo
* cd into repo's root
* Build image `docker build -t ngx_stat_img .`
* Run container `docker run -p 80:80 --name=ngx_stat -d ngx_stat_img`

Last command return container id, so you can use it to check logs by `docker logs _container_id_`.

To check that it is works go to `http://localhost` and other location like `http://localhost/304` or `http://localhost/500`,
and on `http://localhost/show_stat/` to get statistic page or `curl http://localhost/stat` to get JSON.

You also can run container with test suite:

* cd in tests directory
* Build image `docker build -t ngx_test_img .`
* Run container `docker run -i -t --link=ngx_stat:web ngx_test_img`

Last command should return result of tests.

### Get image from docker index

You can just pull image from index.docker.io by `docker pull shoonoise/lua-nginx-statistics`.

## Tips and Tricks

1. You may want to use lua [cjson](http://www.kyne.com.au/~mark/software/lua-cjson-manual.html) module instead of `liblua5.1-json` for performance purpose.
Just install cjson from [ppa](https://launchpad.net/ubuntu/+source/lua-cjson) and replace `local json = require("json")` to `local json = require("cjson")`
in lua scripts.

2. Set `$stat_counter` nginx variable in config to group statuses.

For example set `$stat_counter $uri` to collect statuses for each uri separately.


debian/changelog
lua-nginx-statistics (0.1ppa1) precise; urgency=low

  * Initial release

 -- Alexander Kushnarev <avkushnarev@gmail.com>  Thu, 03 Apr 2014 14:32:31 +0000


tests.py
from urlparse import urljoin
import requests
import unittest

HOST = "http://localhost"


class Base(unittest.TestCase):
    """
    ATTENTION: Test name define a target status
    """

    @staticmethod
    def _get_current_stat(name):
        reply = requests.get(urljoin(HOST, "status_stat"),
                             headers={'content-type': 'application/json'})
        return reply.json().get(name)

    def _check_counter(fn):
        """Make a little magic to avoid code duplication.
           This wrapper generate test based on test method's name """
        def wrapper(fn):
            status = fn._testMethodName.split("_")[-1]
            current = Base._get_current_stat(status)
            requests.get(urljoin(HOST, status), allow_redirects=False)
            new = Base._get_current_stat(status)
            unittest.TestCase.assertEqual(fn, new - current, 1,
                                          "Counter for status %s should be %s, but %s" % (status, current + 1, new))
        return wrapper

    @_check_counter
    def test_get_204(self):
        pass

    @_check_counter
    def test_get_500(self):
        pass

    @_check_counter
    def test_get_404(self):
        pass

    @_check_counter
    def test_get_302(self):
        pass

if __name__ == '__main__':
    unittest.main()


tests/tests.py
from urlparse import urljoin
import requests
import unittest
import os
import json

HOST = "http://%s" % os.environ['WEB_PORT_80_TCP_ADDR']


class Base(unittest.TestCase):

    @staticmethod
    def _get_current_stat(stat_type):
        reply = requests.get(urljoin(HOST, "stat"),
                             headers={'content-type': 'application/json'})
        reply.raise_for_status()
        return json.loads(reply.content)[stat_type]

    def test_statuses(self):
        """Check statuses statistic."""
        locations = ["105", "204", "302", "404", "500"]

        for location in locations:
            current_statuses = self._get_current_stat("status")
            status_class = "%sxx" % location[0]

            if location in current_statuses:
                current_counter = current_statuses[location][status_class]
            else:
                current_counter = 0
            total_counter = current_statuses['_total'][status_class]
            requests.get(urljoin(HOST, '%s' % location), allow_redirects=False)
            new_statuses = self._get_current_stat("status")
            new_counter = new_statuses[location][status_class]
            new_total = new_statuses['_total'][status_class]

            self.assertEqual(new_counter - current_counter, 1)
            self.assertEqual(new_total - total_counter, 1)

    def test_delays(self):
        """Check timings statistic.
           Disclaimer:
           I believe there is no lag on local host.
           So, on some cases you can't trust this tests on 100%"""

        expected_delays = {'0': '0-100',
                           '02': '100-500',
                           '06': '500-1000',
                           '1': '1000-inf'}

        for location, delay in expected_delays.items():
            current_timings = self._get_current_stat("timings")
            if location in current_timings:
                current_counter = current_timings[location][delay]
            else:
                current_counter = 0
            total_counter = current_timings['_total'][delay]
            requests.get(urljoin(HOST, '%s' % location))
            new_timings = self._get_current_stat("timings")
            new_counter = new_timings[location][delay]
            new_total = new_timings['_total'][delay]

            self.assertEqual(new_counter - current_counter, 1)
            self.assertEqual(new_total - total_counter, 1)

if __name__ == '__main__':
    unittest.main()


show_stat.lua
local json = require("json")
local common = require("common_stat")
local stat_dict = ngx.shared.stat_dict

json_data = stat_dict:get("stat")

if json_data then
  data = json.decode(json_data)
else
  data = {["status"]  = {["_total"] = common.init_location_group("status")},
          ["timings"] = {["_total"] = common.init_location_group("timings")}}
end

-- Build _total field
local function build_total_field(stat_type)
  local _total = common.init_location_group(stat_type)
  ngx.log(ngx.INFO, "Total ", json.encode(_total))
  for _, v in pairs(data[stat_type]) do
    for s, c in pairs(v) do
      _total[s] = _total[s] + c
    end
  end
  return _total
end

data["status"]["_total"] = build_total_field("status")
data["timings"]["_total"] = build_total_field("timings")

ngx.header.content_type = "application/json"
ngx.say(json.encode(data))


collect_statuses.lua
module ("collect_statuses", package.seeall)
local json = require("json")
local stat_dict = ngx.shared.stat_dict
local status = tostring(ngx.status)
local location_group = ngx.var.stat_counter

function init_location_group()
  local location_group = {
                          ["1xx"] = 0,
                          ["2xx"] = 0,
                          ["3xx"] = 0,
                          ["4xx"] = 0,
                          ["5xx"] = 0,
                          ["xxx"] = 0
                         }
  return location_group
end

-- Set default group, if it's not defined by nginx variable
if not location_group or location_group == "" then
  location_group = 'other_locations'
end

-- Get serialized data from dict or create new
local ser_stat = stat_dict:get("stat")
if ser_stat then
  stat = json.decode(ser_stat)
else
  stat = {["status"] = {[location_group] = init_location_group()}}
end

function get_status_code_class(status_code)
  if     status_code:sub(1,1) == '1' then return "1xx"
  elseif status_code:sub(1,1) == '2' then return "2xx"
  elseif status_code:sub(1,1) == '3' then return "3xx"
  elseif status_code:sub(1,1) == '4' then return "4xx"
  elseif status_code:sub(1,1) == '5' then return "5xx"
  else                                    return "xxx"
  end
end

for i, _ in pairs(stat) do
  if i == "status" then
    local status_class = get_status_code_class(status)

    if not stat["status"][location_group] then
      stat["status"][location_group] = init_location_group()
    end

    if stat["status"][location_group][status_class] then
      local old_count = stat["status"][location_group][status_class]
      stat["status"][location_group][status_class] = old_count + 1
    end
  end
end

stat_dict:set("stat", json.encode(stat))
ngx.log(ngx.INFO, " stat ", json.encode(stat))


show_statuses_stat.lua
local json = require "json"
local stat_dict = ngx.shared.stat_dict
local collect_module = require "collect_statuses"

ser_data = stat_dict:get("stat")

if not ser_data then
  ngx.say("No data yet.")
  ngx.exit(ngx.OK)
end

data = json.decode(ser_data)

-- Build _total field
local _total = collect_module.init_location_group()
for _, v in pairs(data["status"]) do
  for s, c in pairs(v) do
    _total[s] = _total[s] + c
  end
end
data["status"]["_total"] = _total

-- Build reply

if ngx.req.get_headers()["Content-Type"] == "application/json" then
  -- Reply JSON if content type application/json
  ngx.header.content_type = "application/json"
  ngx.say(json.encode(data))

else
  -- Reply HTML in other ways
  local html_reply = {}

  table.insert(html_reply, "<html><head><title>Nginx statistic</title></head><body>")
  table.insert(html_reply, "<h1>Statuses:</h1>")

  for group, value in pairs(data["status"]) do
    table.insert(html_reply, "<h2>" .. group .. "</h2")
    for s, c in pairs(value) do
      table.insert(html_reply, "<b>" .. s .. "</b> has occurred "
                               .. tostring(c) .. " times" .. "</p>")
    end
  end

  table.insert(html_reply, "</body></html>")

  ngx.say(html_reply)
end


static/js/show_stat.js
/**
 * Created by Alexander Kushnarev on 26.03.14.
 */

$( document ).ready(function() {
    $.getJSON( "/stat", function( data ) {
        var statistic = function () {
            var main = "<h1>Nginx statistic page</h1>";
            $.each(data, function (stat_type, stat_data) {
                main += "<h2>" + stat_type + "</h2>";
                $.each(stat_data, function (loc, values) {
                    var counters_list = '<h3> In '+ loc +':</h3>';
                    counters_list += "<ul>";
                    $.each(values, function (s, count) {
                        counters_list += "<li>" + "<b>" + s + "</b>" +" status was occurred " + count + " times " + "</li>";
                    });
                    counters_list += "</ul>";
                    main += counters_list;
                })
            });
            return main;
        };
        $("body").append(statistic);
    });
});


collect_stats.lua
local json = require("json")
local common = require("common_stat")

local stat_dict = ngx.shared.stat_dict
local status = tostring(ngx.status)
local upstream_time = tonumber(ngx.var.request_time) * 1000
local location_group = ngx.var.stat_counter

-- Set default group, if it's not defined by nginx variable
if not location_group or location_group == "" then
  location_group = 'other_locations'
end

json_stat = stat_dict:get("stat")
if json_stat then
  ngx.log(ngx.INFO, "There is serialized data ", json_stat)
  stat_data = json.decode(json_stat)
else
  ngx.log(ngx.INFO, " There is no data in shared dict ")
  stat_data = {["status"] = {[location_group] = common.init_location_group("status")},
               ["timings"] = {[location_group] = common.init_location_group("timings")}}
end

-- Get serialized data from shared dict or create new if it doesn exist
local function get_data(stat_type)
  if stat_data[stat_type][location_group] then
    ngx.log(ngx.INFO, " There is data for ", stat_type, " type.")
    if stat_data[stat_type][location_group] then
      return stat_data[stat_type][location_group]
    else
      return init_location_group(stat_type)
    end
  else
    ngx.log(ngx.INFO, " There is no data for ", stat_type, " type. So init it.")
    return common.init_location_group(stat_type)
  end
end

-- Define in which category status should be
local function get_status_code_class(status_code)
  ngx.log(ngx.INFO, "Get code class for ", status_code)
  if     status_code:sub(1,1) == '1' then return "1xx"
  elseif status_code:sub(1,1) == '2' then return "2xx"
  elseif status_code:sub(1,1) == '3' then return "3xx"
  elseif status_code:sub(1,1) == '4' then return "4xx"
  elseif status_code:sub(1,1) == '5' then return "5xx"
  else                                    return "xxx"
  end
end

-- Define in which category response time should be
local function get_histogram_state(t)
  ngx.log(ngx.INFO, "Get histogram state for ", t)
  if not t                     then return "undef"
  elseif t >= 0  and t < 100  then return "0-100"
  elseif t >= 100 and t < 500  then return "100-500"
  elseif t >= 500 and t < 1000 then return "500-1000"
  elseif t >= 1000             then return "1000-inf"
  end
end

local function build_stat(location_data, stat_type, value)
  ngx.log(ngx.INFO, " Build stat for ", stat_type, "type. And ", value, " value.")
  ngx.log(ngx.INFO, "Current location data ", json.encode(location_data))
  if stat_type == "status" then
    ngx.log(ngx.INFO, " Get class for status ")
    stat_class = get_status_code_class(value)
  elseif stat_type == "timings" then
    ngx.log(ngx.INFO, " Get class for timings ")
    stat_class = get_histogram_state(value)
  end
  ngx.log(ngx.INFO, " Class ", stat_class)

  location_data[stat_class] = location_data[stat_class] + 1
  ngx.log(ngx.INFO, " New location data ", json.encode(location_data))
  return location_data
end

local statuses = build_stat(get_data("status"), "status", status)
ngx.log(ngx.INFO, " statuses ", json.encode(statuses))

local timings = build_stat(get_data("timings"), "timings", upstream_time)
ngx.log(ngx.INFO, " timings ", json.encode(timings))

stat_data["status"][location_group] = statuses
stat_data["timings"][location_group] = timings

ngx.log(ngx.INFO, " stat data ", json.encode(stat_data))

stat_dict:set("stat", json.encode(stat_data))
